R/colsums_rowsums_linter.R

```diff
@@ -0,0 +1,156 @@
+  "
+
+  xpath <- glue::glue("{sums_xpath} | {means_xpath}")
+
+  # This doesn't handle the case when MARGIN and FUN are named and in a different position
+  # but this should be relatively rate
+  var_xpath  <- "expr[position() = 2]"
+  margin_xpath <- "expr[position() = 3]"
+  fun_xpath <- "expr[position() = 4]"
+
+  craft_msg <- function(var, margin, fun, narm_val) {
+
+    if (is.na(xml2::xml_find_first(margin, "OP-COLON"))) {
+      l1 <- xml2::xml_text(margin)
+      l2 <- NULL
+    } else {
+      l1 <- xml2::xml_text(xml2::xml_find_first(margin, "expr[1]"))
+      l2 <- xml2::xml_text(xml2::xml_find_first(margin, "expr[2]"))
+    }
+
+    # See #1764 for details about various cases. In short:
+    # - If apply(., 1:l2, sum) -> rowSums(., dims = l2)
+    # - If apply(., l1:l2, sum) -> rowSums(colSums(., dims = l1 - 1), dims = l2 - l1 + 1)
+    # - This last case can be simplified to a simple colSums() call if l2 = length(dim(.))
+    # - dims argument can be dropped if equals to 1. This notably is the case for matrices
+    if (is.null(l2)) {
+      l2 <- l1
+    }
+
+    l1 <- suppressWarnings(as.integer(re_substitutes(l1, "L$", "")))
+    l2 <- suppressWarnings(as.integer(re_substitutes(l2, "L$", "")))
+
+    if (!is.na(narm_val)) {
+      narm <- glue::glue(", na.rm = {narm_val}")
+    } else {
+      narm <- ""
+    }
+
+    if (identical(l1, 1L)) {
+      reco <- glue::glue("row{fun}s({var}{narm}, dims = {l2})")
+    } else if (anyNA(c(l1, l2))) {
+      # Return generic error messages if l1 and l2 can't be parsed since we need
+      # to do arithmetic operations on them to produce custom messages
+      reco <- glue::glue(
+        "row{fun}s(col{fun}s({var}{narm}, dims = l1 - 1), dims = l2 - l1 + 1)",
+        " or ",
+        "row{fun}s({var}{narm}, dims = l2) if l1 == 1L",
+        " or ",
+        "col{fun}s({var}{narm}, dims = l1 - 1) if {var} has l2 dimensions"
+      )
+    } else {
+      reco <- glue::glue(
+        "row{fun}s(col{fun}s({var}{narm}, dims = {l1 - 1}), dims = {l2 - l1 + 1})",
+        " or ",
+        "col{fun}s({var}{narm}, dims = {l1 - 1}) if {var} has {l2} dimensions"
+      )
+    }
+
+    # It's easier to remove this after the fact, rather than having never ending if/elses
+    reco <- gsub(", dims = 1", "", reco, fixed = TRUE)
+
+    return(reco)
+  }
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "expression")) {
+      return(list())
+    }
+    xml <- source_expression$xml_parsed_content
+
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+
+    var <- xml2::xml_text(xml2::xml_find_all(bad_expr, var_xpath))
+
+    fun <- xml2::xml_text(xml2::xml_find_all(bad_expr, fun_xpath))
+    fun <- tools::toTitleCase(fun)
+
+    margin <- xml2::xml_find_all(bad_expr, margin_xpath)
+
+    narm_val <- xml2::xml_text(
+      xml2::xml_find_first(bad_expr, "SYMBOL_SUB[text() = 'na.rm']/following-sibling::expr")
+    )
+
+    recos <- lapply(seq_along(bad_expr), function(i) {
```

Comment:
This could be a `mapply()` call instead.

Reply:
Don't we prefer `Map()` over `mapply()`?

https://github.com/r-lib/lintr/blob/08098ef77adb1eccdf18f8b44a7053f36297df28/R/zzz.R#L66

Reply:
Good catch. Still should be `Map()` then.